File: lib/screens/habit/habit_stats_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:fl_chart/fl_chart.dart';
import '../../data/habit/habit_models.dart';
import '../../l10n/app_localizations.dart';
import '../../providers/habit_provider.dart';
import '../../theme/app_theme.dart';
import '../../utils/datetime/date_utils.dart';
import '../../utils/stats/stats_utils.dart';
import '../../utils/stats/streak_calculator.dart';
import '../../widgets/stats/compact_vertical_stat_card.dart';

import '../../widgets/common/week_navigator.dart';

enum DayTimeName { morning, afternoon, evening, night }

class HabitStatsScreen extends StatefulWidget {
  const HabitStatsScreen({super.key});

  @override
  State<HabitStatsScreen> createState() => _HabitStatsScreenState();
}

class _HabitStatsScreenState extends State<HabitStatsScreen> {
  DateTime _focusedDate = DateTime.now();

  @override
  Widget build(BuildContext context) {
    final habits = context.watch<HabitProvider>().habits;
    final l10n = AppLocalizations.of(context)!;

    final startOfWeek = AppDateUtils.startOfWeekMonday(_focusedDate);
    final endOfWeek = startOfWeek.add(
      const Duration(days: 6, hours: 23, minutes: 59, seconds: 59),
    );
    final bool isCurrentRealWeek = AppDateUtils.isWithinInclusive(
      DateTime.now(),
      startOfWeek,
      endOfWeek,
    );

    // KPI Calculations
    final int totalCompletions = habits.fold(
      0,
      (sum, h) => sum + h.completions.length,
    );
    int allTimeBestStreak = 0;
    for (var h in habits) {
      final int s = StreakCalculator.calculateBestStreak(h.completions);
      if (s > allTimeBestStreak) allTimeBestStreak = s;
    }

    return Scaffold(
      body: habits.isEmpty
          ? Center(child: Text(l10n.habitStatsLabelEmpty))
          : ListView(
              padding: AppTheme.paddingScreenVertical,
              children: [
                // --- WEEK NAVIGATOR ---
                WeekNavigator(
                  focusedDate: _focusedDate,
                  onShiftWeek: _changeWeek,
                ),
                const SizedBox(height: AppTheme.spaceLarge),

                // --- COMPACT KPIs ---
                Wrap(
                  alignment: WrapAlignment.center,
                  spacing: AppTheme.spaceMedium,
                  runSpacing: AppTheme.spaceMedium,
                  children: [
                    CompactVerticalStatCard(
                      title: l10n.habitStatsLabelTotal,
                      value: "$totalCompletions",
                      icon: Icons.timeline,
                      color: AppTheme.habitColor,
                    ),
                    CompactVerticalStatCard(
                      title: l10n.habitStatsLabelActive,
                      value: "${habits.length}",
                      icon: Icons.all_inclusive,
                      color: AppTheme.successColor,
                    ),
                    CompactVerticalStatCard(
                      title: l10n.habitStatsLabelBestStreak,
                      value: "$allTimeBestStreak",
                      icon: Icons.military_tech,
                      color: AppTheme.warningColor,
                    ),
                  ],
                ),
                const SizedBox(height: AppTheme.space3XLarge),

                // --- WEEKLY VOLUME CHART ---
                Text(
                  l10n.habitStatsLabelVelocity,
                  style: const TextStyle(
                    fontSize: AppTheme.fsHeadingSmall,
                    fontWeight: AppTheme.fwBold,
                  ),
                ),
                Text(
                  l10n.habitStatsLabelCompletionsDaily,
                  style: const TextStyle(
                    fontSize: AppTheme.fsBodySmall,
                    color: Colors.grey,
                  ),
                ),
                const SizedBox(height: AppTheme.spaceLarge),
                SizedBox(
                  height: 200,
                  child: _buildBarChart(
                    habits,
                    startOfWeek,
                    endOfWeek,
                    isCurrentRealWeek,
                  ),
                ),
                const SizedBox(height: AppTheme.space3XLarge),

                // --- TIME OF DAY BREAKDOWN ---
                _buildTimeOfDayBreakdownCard(habits),
                const SizedBox(height: AppTheme.space3XLarge),

                // --- LEADERBOARD ---
                Text(
                  l10n.habitStatsLabelCurrentStreak,
                  style: const TextStyle(
                    fontSize: AppTheme.fsHeadingSmall,
                    fontWeight: AppTheme.fwBold,
                  ),
                ),
                const SizedBox(height: AppTheme.spaceMedium),
                ...habits.map((habit) {
                  final int currentStreak =
                      StreakCalculator.calculateCurrentStreak(
                        habit.completions,
                      );
                  final int bestStreak = StreakCalculator.calculateBestStreak(
                    habit.completions,
                  );

                  return Card(
                    margin: const EdgeInsets.only(bottom: AppTheme.spaceSmall),
                    shape: const RoundedRectangleBorder(
                      side: BorderSide(
                        color: AppTheme.habitColor,
                        width: AppTheme.neoBorderWidth,
                      ),
                      borderRadius: AppTheme.brNeo,
                    ),
                    child: ListTile(
                      leading: const Icon(
                        Icons.stars,
                        color: AppTheme.habitColor,
                      ),
                      title: Text(
                        habit.name,
                        style: const TextStyle(fontWeight: AppTheme.fwBold),
                      ),
                      trailing: Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        crossAxisAlignment: CrossAxisAlignment.end,
                        children: [
                          Row(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              Text(
                                "$currentStreak",
                                style: const TextStyle(
                                  fontSize: 16,
                                  fontWeight: AppTheme.fwExtraBold,
                                  color: AppTheme.warningColor,
                                ),
                              ),
                              const SizedBox(width: 4),
                              const Icon(
                                Icons.whatshot,
                                size: 16,
                                color: AppTheme.warningColor,
                              ),
                            ],
                          ),
                          Text(
                            l10n.habitStatsLabelBest(bestStreak),
                            style: const TextStyle(
                              fontSize: 10,
                              color: Colors.grey,
                            ),
                          ),
                        ],
                      ),
                    ),
                  );
                }),
                const SizedBox(height: AppTheme.space3XLarge),
              ],
            ),
    );
  }

  // --- CHART BUILDERS ---
  Widget _buildBarChart(
    List<Habit> habits,
    DateTime startOfWeek,
    DateTime endOfWeek,
    bool isCurrentRealWeek,
  ) {
    final l10n = AppLocalizations.of(context)!;
    final dailyCounts = _getCompletionsForWeek(habits, startOfWeek, endOfWeek);
    final todayIndex = DateTime.now().weekday - 1;
    final weekdays = [
      l10n.commonWeekdayMon,
      l10n.commonWeekdayTue,
      l10n.commonWeekdayWed,
      l10n.commonWeekdayThu,
      l10n.commonWeekdayFri,
      l10n.commonWeekdaySat,
      l10n.commonWeekdaySun,
    ];

    final maxY = StatsUtils.maxValue(dailyCounts, minimum: 5);

    return BarChart(
      BarChartData(
        alignment: BarChartAlignment.spaceAround,
        maxY: maxY + 1,
        barTouchData: const BarTouchData(enabled: false),
        titlesData: FlTitlesData(
          bottomTitles: AxisTitles(
            sideTitles: SideTitles(
              showTitles: true,
              getTitlesWidget: (double value, TitleMeta meta) {
                final int dayIndex = value.toInt();
                final bool isToday =
                    isCurrentRealWeek && (dayIndex == todayIndex);
                return Padding(
                  padding: const EdgeInsets.only(top: 8.0),
                  child: Text(
                    weekdays[dayIndex],
                    style: TextStyle(
                      fontSize: AppTheme.fsBodySmall,
                      fontWeight: isToday ? AppTheme.fwBold : AppTheme.fwNormal,
                      color: isToday ? AppTheme.habitColor : Colors.grey,
                    ),
                  ),
                );
              },
            ),
          ),
          leftTitles: const AxisTitles(),
          topTitles: const AxisTitles(),
          rightTitles: const AxisTitles(),
        ),
        gridData: const FlGridData(show: false),
        borderData: FlBorderData(show: false),
        barGroups: List.generate(7, (index) {
          final bool isToday = isCurrentRealWeek && (index == todayIndex);
          return BarChartGroupData(
            x: index,
            barRods: [
              BarChartRodData(
                toY: dailyCounts[index].toDouble(),
                color: isToday
                    ? AppTheme.habitColor
                    : AppTheme.habitColor.withValues(
                        alpha: AppTheme.opacityMedium,
                      ),
                width: 16,
                borderRadius: BorderRadius.circular(4),
                backDrawRodData: BackgroundBarChartRodData(
                  show: true,
                  toY: maxY + 1,
                  color: Colors.grey.withValues(alpha: 0.1),
                ),
              ),
            ],
          );
        }),
      ),
    );
  }

  void _changeWeek(int days) {
    setState(() {
      _focusedDate = _focusedDate.add(Duration(days: days));
    });
  }

  // --- DATA PROCESSING ---
  List<int> _getCompletionsForWeek(
    List<Habit> habits,
    DateTime startOfWeek,
    DateTime endOfWeek,
  ) {
    final List<int> dailyCounts = List.filled(7, 0);
    for (var habit in habits) {
      for (var c in habit.completions) {
        if (c.isAfter(startOfWeek.subtract(const Duration(seconds: 1))) &&
            c.isBefore(endOfWeek)) {
          dailyCounts[c.weekday - 1]++;
        }
      }
    }
    return dailyCounts;
  }

  Map<DayTimeName, int> _getTimeOfDayData(List<Habit> habits) {
    int morning = 0; // 5 AM - 12 PM
    int afternoon = 0; // 12 PM - 5 PM
    int evening = 0; // 5 PM - 9 PM
    int night = 0; // 9 PM - 5 AM

    for (var habit in habits) {
      for (var c in habit.completions) {
        if (c.hour >= 5 && c.hour < 12) {
          morning++;
        } else if (c.hour >= 12 && c.hour < 17) {
          afternoon++;
        } else if (c.hour >= 17 && c.hour < 21) {
          evening++;
        } else {
          night++;
        }
      }
    }
    return {
      DayTimeName.morning: morning,
      DayTimeName.afternoon: afternoon,
      DayTimeName.evening: evening,
      DayTimeName.night: night,
    };
  }

  String _formatHabitCount(int count) {
    if (count == 1) return '1 completion';
    return '$count completions';
  }

  Widget _buildTimeOfDayBreakdownCard(List<Habit> habits) {
    final data = _getTimeOfDayData(habits);
    final total = data.values.fold<int>(0, (sum, value) => sum + value);
    final l10n = AppLocalizations.of(context)!;

    final buckets = [
      TimeOfDayBucket(
        label: l10n.commonDaytimeMorning,
        subtitle: l10n.commonDaytimeMorningRange,
        count: data[DayTimeName.morning] ?? 0,
        color: AppTheme.warningColor,
        icon: Icons.wb_sunny_outlined,
      ),
      TimeOfDayBucket(
        label: l10n.commonDaytimeAfternoon,
        subtitle: l10n.commonDaytimeAfternoonRange,
        count: data[DayTimeName.afternoon] ?? 0,
        color: AppTheme.taskColor,
        icon: Icons.sunny,
      ),
      TimeOfDayBucket(
        label: l10n.commonDaytimeEvening,
        subtitle: l10n.commonDaytimeEveningRange,
        count: data[DayTimeName.evening] ?? 0,
        color: AppTheme.habitColor,
        icon: Icons.nightlight_round,
      ),
      TimeOfDayBucket(
        label: l10n.commonDaytimeNight,
        subtitle: l10n.commonDaytimeNightRange,
        count: data[DayTimeName.night] ?? 0,
        color: AppTheme.userColor,
        icon: Icons.nights_stay_outlined,
      ),
    ];

    // Removed Card and inner background Container to unify layout styles
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Row(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    l10n.habitStatsLabelCompletionsTime,
                    style: const TextStyle(
                      fontSize: AppTheme.fsHeadingSmall,
                      fontWeight: AppTheme.fwBold,
                    ),
                  ),
                  const SizedBox(height: 4),
                  Text(
                    l10n.habitStatsLabelCompletionsDistribution,
                    style: const TextStyle(
                      fontSize: AppTheme.fsBodySmall,
                      color: Colors.grey,
                    ),
                  ),
                ],
              ),
            ),
            const SizedBox(width: 16),
            Text(
              _formatHabitCount(total),
              style: const TextStyle(
                fontSize: 15,
                fontWeight: AppTheme.fwExtraBold,
              ),
            ),
          ],
        ),
        const SizedBox(height: 18),
        ClipRRect(
          borderRadius: BorderRadius.circular(999),
          child: Container(
            height: 16,
            color: Theme.of(context).colorScheme.surfaceContainerHighest,
            child: total == 0
                ? Container(color: Colors.grey.withValues(alpha: 0.12))
                : Row(
                    children: buckets
                        .where((bucket) => bucket.count > 0)
                        .map(
                          (bucket) => Expanded(
                            flex: bucket.count,
                            child: Container(color: bucket.color),
                          ),
                        )
                        .toList(),
                  ),
          ),
        ),
        const SizedBox(height: 18),
        Wrap(
          spacing: 12,
          runSpacing: 12,
          children: buckets.map((bucket) {
            final percent = total == 0
                ? 0
                : ((bucket.count / total) * 100).round();
            return SizedBox(
              width: MediaQuery.of(context).size.width < 420
                  ? double.infinity
                  : (MediaQuery.of(context).size.width - 56) / 2,
              child: Container(
                padding: const EdgeInsets.all(14),
                decoration: BoxDecoration(
                  color: bucket.color.withValues(alpha: 0.08),
                  borderRadius: BorderRadius.circular(18),
                  border: Border.all(
                    color: bucket.color.withValues(alpha: 0.22),
                  ),
                ),
                child: Row(
                  children: [
                    Container(
                      width: 36,
                      height: 36,
                      decoration: BoxDecoration(
                        color: bucket.color.withValues(alpha: 0.14),
                        shape: BoxShape.circle,
                      ),
                      child: Icon(bucket.icon, color: bucket.color, size: 20),
                    ),
                    const SizedBox(width: 12),
                    Expanded(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Text(
                            bucket.label,
                            style: const TextStyle(
                              fontSize: 14,
                              fontWeight: AppTheme.fwBold,
                              height: 1.2,
                            ),
                          ),
                          const SizedBox(height: 2),
                          Text(
                            bucket.subtitle,
                            style: TextStyle(
                              fontSize: 11,
                              color: Theme.of(
                                context,
                              ).colorScheme.onSurfaceVariant,
                            ),
                          ),
                        ],
                      ),
                    ),
                    Column(
                      crossAxisAlignment: CrossAxisAlignment.end,
                      children: [
                        Text(
                          '${bucket.count}',
                          style: const TextStyle(
                            fontSize: 16,
                            fontWeight: AppTheme.fwExtraBold,
                          ),
                        ),
                        Text(
                          '$percent%',
                          style: TextStyle(
                            fontSize: 11,
                            color: Theme.of(
                              context,
                            ).colorScheme.onSurfaceVariant,
                          ),
                        ),
                      ],
                    ),
                  ],
                ),
              ),
            );
          }).toList(),
        ),
      ],
    );
  }
}

File: app/src/main/java/io/github/benji377/timety/ui/screens/habit/HabitStatsScreen.kt
package io.github.benji377.timety.ui.screens.habit

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AllInclusive
import androidx.compose.material.icons.filled.MilitaryTech
import androidx.compose.material.icons.filled.NightlightRound
import androidx.compose.material.icons.filled.Stars
import androidx.compose.material.icons.filled.Timeline
import androidx.compose.material.icons.filled.WbSunny
import androidx.compose.material.icons.filled.Whatshot
import androidx.compose.material.icons.outlined.NightsStay
import androidx.compose.material.icons.outlined.WbSunny
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import io.github.benji377.timety.R
import io.github.benji377.timety.data.model.habit.HabitCompletionEntity
import io.github.benji377.timety.data.model.habit.HabitWithCompletions
import io.github.benji377.timety.ui.components.common.AccentBadge
import io.github.benji377.timety.ui.components.common.NeoCard
import io.github.benji377.timety.ui.components.common.WeekNavigator
import io.github.benji377.timety.ui.components.stats.StatCard
import io.github.benji377.timety.ui.components.stats.StatCardStyle
import io.github.benji377.timety.ui.theme.AppTheme
import io.github.benji377.timety.ui.theme.HabitColor
import io.github.benji377.timety.ui.theme.SuccessColor
import io.github.benji377.timety.ui.theme.TaskColor
import io.github.benji377.timety.ui.theme.UserColor
import io.github.benji377.timety.ui.theme.WarningColor
import io.github.benji377.timety.ui.utils.quantityString
import io.github.benji377.timety.ui.viewmodel.AppViewModelProvider
import io.github.benji377.timety.ui.viewmodel.HabitViewModel
import io.github.benji377.timety.util.datetime.AppDateUtils
import io.github.benji377.timety.util.stats.StatsUtils
import io.github.benji377.timety.util.stats.StreakCalculator
import java.time.LocalDate
import java.time.ZoneId
import kotlin.math.roundToInt


/**
 * Habit statistics: weekly summary cards, a daily completion chart, a time-of-day breakdown, and
 * per-habit streaks. Streaks are computed all-time; the other sections follow the selected week.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun HabitStatsScreen(
    viewModel: HabitViewModel = viewModel(factory = AppViewModelProvider.Factory)
) {
    val habits by viewModel.habitsWithCompletions.collectAsState()
    var focusedDate by remember { mutableStateOf(LocalDate.now()) }

    Scaffold { paddingValues ->
        if (habits.isEmpty()) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(paddingValues),
                contentAlignment = Alignment.Center,
            ) {
                Text(stringResource(R.string.habitStatsLabelEmpty))
            }
        } else {
            // The summary cards and the time-of-day breakdown follow the week selector, so
            // shifting the week visibly updates the stats. Streaks stay all-time: a
            // per-week streak is meaningless.
            val startOfWeek = AppDateUtils.startOfWeekMonday(focusedDate)
            val endOfWeek = startOfWeek.plusDays(6)
            val zone = ZoneId.systemDefault()
            val weekCompletions = habits.flatMap { hwc ->
                hwc.completions.filter { c ->
                    AppDateUtils.isWithinInclusive(
                        c.completionDate.atZone(zone).toLocalDate(),
                        startOfWeek,
                        endOfWeek
                    )
                }
            }
            val allTimeBestStreak = habits.maxOfOrNull { hwc ->
                StreakCalculator.calculateBestStreak(completionDates(hwc))
            } ?: 0

            LazyColumn(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(paddingValues)
                    .padding(horizontal = AppTheme.spaceLarge),
                contentPadding = PaddingValues(vertical = AppTheme.spaceLarge),
            ) {
                item {
                    WeekNavigator(
                        focusedDate = focusedDate,
                        onShiftWeek = { days -> focusedDate = focusedDate.plusDays(days.toLong()) },
                    )
                    Spacer(modifier = Modifier.height(AppTheme.spaceLarge))
                }

                item {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.spacedBy(
                            AppTheme.spaceMedium,
                            Alignment.CenterHorizontally
                        ),
                    ) {
                        StatCard(
                            title = stringResource(R.string.habitStatsLabelTotal),
                            value = weekCompletions.size.toString(),
                            icon = Icons.Filled.Timeline,
                            color = HabitColor,
                            style = StatCardStyle.COMPACT_VERTICAL,
                        )
                        StatCard(
                            title = stringResource(R.string.habitStatsLabelActive),
                            value = habits.size.toString(),
                            icon = Icons.Filled.AllInclusive,
                            color = SuccessColor,
                            style = StatCardStyle.COMPACT_VERTICAL,
                        )
                        StatCard(
                            title = stringResource(R.string.habitStatsLabelBestStreak),
                            value = allTimeBestStreak.toString(),
                            icon = Icons.Filled.MilitaryTech,
                            color = WarningColor,
                            style = StatCardStyle.COMPACT_VERTICAL,
                        )
                    }
                    Spacer(modifier = Modifier.height(AppTheme.space3XLarge))
                }

                item {
                    Text(
                        text = stringResource(R.string.habitStatsLabelVelocity),
                        fontSize = AppTheme.fsHeadingSmall,
                        fontWeight = AppTheme.fwBold,
                    )
                    Text(
                        text = stringResource(R.string.habitStatsLabelCompletionsDaily),
                        fontSize = AppTheme.fsBodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                    Spacer(modifier = Modifier.height(AppTheme.spaceLarge))
                    HabitVelocityChart(habits = habits, focusedDate = focusedDate)
                    Spacer(modifier = Modifier.height(AppTheme.space3XLarge))
                }

                item {
                    TimeOfDayBreakdownCard(weekCompletions)
                    Spacer(modifier = Modifier.height(AppTheme.space3XLarge))
                }

                item {
                    Text(
                        text = stringResource(R.string.habitStatsLabelCurrentStreak),
                        fontSize = AppTheme.fsHeadingSmall,
                        fontWeight = AppTheme.fwBold,
                    )
                    Spacer(modifier = Modifier.height(AppTheme.spaceMedium))
                }

                items(habits, key = { it.habit.id }) { hwc ->
                    val dates = completionDates(hwc)
                    val currentStreak = StreakCalculator.calculateCurrentStreak(dates)
                    val bestStreak = StreakCalculator.calculateBestStreak(dates)

                    NeoCard(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(bottom = AppTheme.spaceSmall),
                        borderColor = HabitColor,
                    ) {
                        Row(
                            modifier = Modifier.padding(AppTheme.spaceLarge),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Icon(Icons.Filled.Stars, contentDescription = null, tint = HabitColor)
                            Spacer(modifier = Modifier.width(AppTheme.spaceLarge))
                            Text(
                                text = hwc.habit.name,
                                fontWeight = AppTheme.fwBold,
                                modifier = Modifier.weight(1f),
                            )
                            Column(horizontalAlignment = Alignment.End) {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Text(
                                        text = currentStreak.toString(),
                                        fontSize = 16.sp,
                                        fontWeight = AppTheme.fwExtraBold,
                                        color = WarningColor,
                                    )
                                    Spacer(modifier = Modifier.width(4.dp))
                                    Icon(
                                        Icons.Filled.Whatshot,
                                        contentDescription = null,
                                        tint = WarningColor,
                                        modifier = Modifier.size(16.dp),
                                    )
                                }
                                Text(
                                    text = stringResource(R.string.habitStatsLabelBest, bestStreak),
                                    fontSize = AppTheme.fsCaption,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                                )
                            }
                        }
                    }
                }

                item { Spacer(modifier = Modifier.height(AppTheme.space3XLarge)) }
            }
        }
    }
}

private fun completionDates(hwc: HabitWithCompletions): List<LocalDate> =
    hwc.completions.map { it.completionDate.atZone(ZoneId.systemDefault()).toLocalDate() }

@Composable
private fun HabitVelocityChart(habits: List<HabitWithCompletions>, focusedDate: LocalDate) {
    val startOfWeek = AppDateUtils.startOfWeekMonday(focusedDate)
    val endOfWeek = startOfWeek.plusDays(6)
    val today = LocalDate.now()
    val isCurrentRealWeek = AppDateUtils.isWithinInclusive(today, startOfWeek, endOfWeek)
    val todayIndex = today.dayOfWeek.value - 1

    val dailyCounts = IntArray(7)
    val zone = ZoneId.systemDefault()
    habits.forEach { hwc ->
        hwc.completions.forEach { c ->
            val date = c.completionDate.atZone(zone).toLocalDate()
            if (!date.isBefore(startOfWeek) && !date.isAfter(endOfWeek)) {
                dailyCounts[date.dayOfWeek.value - 1]++
            }
        }
    }

    val maxY = StatsUtils.maxValue(dailyCounts.toList(), minimum = 5.0)
    val weekdays = listOf(
        stringResource(R.string.commonWeekdayMon),
        stringResource(R.string.commonWeekdayTue),
        stringResource(R.string.commonWeekdayWed),
        stringResource(R.string.commonWeekdayThu),
        stringResource(R.string.commonWeekdayFri),
        stringResource(R.string.commonWeekdaySat),
        stringResource(R.string.commonWeekdaySun),
    )

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(200.dp),
        horizontalArrangement = Arrangement.SpaceAround,
        verticalAlignment = Alignment.Bottom,
    ) {
        for (i in 0 until 7) {
            val isToday = isCurrentRealWeek && i == todayIndex
            val fraction = (dailyCounts[i] / (maxY + 1)).toFloat().coerceIn(0f, 1f)
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Bottom,
                modifier = Modifier
                    .weight(1f)
                    .fillMaxHeight(),
            ) {
                Box(
                    modifier = Modifier
                        .weight(1f)
                        .fillMaxWidth(0.5f)
                        // Solid neutral track (was an alpha-tinted onSurfaceVariant) - the bar
                        // chart's gutter is chrome around the data, not itself a data mark.
                        .background(
                            MaterialTheme.colorScheme.surfaceVariant,
                            RoundedCornerShape(4.dp)
                        ),
                    contentAlignment = Alignment.BottomCenter,
                ) {
                    // Flat, solid bar for every day - fading the non-today bars was a soft alpha
                    // fill standing in for a boolean emphasis, not an encoded data value, so it's
                    // gone. Today still needs a solid affordance beyond the label (bold text alone
                    // is too weak a signal at this bar size), so today's bar gets a solid outline
                    // border instead - a solid differentiator, not restored alpha.
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .fillMaxHeight(fraction)
                            .background(
                                color = HabitColor,
                                shape = RoundedCornerShape(4.dp),
                            )
                            .then(
                                if (isToday) {
                                    Modifier.border(
                                        AppTheme.borderHairline,
                                        MaterialTheme.colorScheme.outline,
                                        RoundedCornerShape(4.dp),
                                    )
                                } else {
                                    Modifier
                                }
                            ),
                    )
                }
                Spacer(modifier = Modifier.height(AppTheme.spaceSmall))
                Text(
                    text = weekdays[i],
                    fontSize = AppTheme.fsBodySmall,
                    fontWeight = if (isToday) AppTheme.fwBold else AppTheme.fwNormal,
                    color = if (isToday) HabitColor else MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }
    }
}

private data class TimeOfDayBucket(
    val label: String,
    val subtitle: String,
    val count: Int,
    val color: Color,
    val icon: ImageVector,
)

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun TimeOfDayBreakdownCard(completions: List<HabitCompletionEntity>) {
    val zone = ZoneId.systemDefault()
    var morning = 0
    var afternoon = 0
    var evening = 0
    var night = 0
    completions.forEach { c ->
        val hour = c.completionDate.atZone(zone).hour
        when (hour) {
            in 5..11 -> morning++
            in 12..16 -> afternoon++
            in 17..20 -> evening++
            else -> night++
        }
    }
    val total = morning + afternoon + evening + night

    val buckets = listOf(
        TimeOfDayBucket(
            label = stringResource(R.string.commonDaytimeMorning),
            subtitle = stringResource(R.string.commonDaytimeMorningRange),
            count = morning,
            color = WarningColor,
            icon = Icons.Outlined.WbSunny,
        ),
        TimeOfDayBucket(
            label = stringResource(R.string.commonDaytimeAfternoon),
            subtitle = stringResource(R.string.commonDaytimeAfternoonRange),
            count = afternoon,
            color = TaskColor,
            icon = Icons.Filled.WbSunny,
        ),
        TimeOfDayBucket(
            label = stringResource(R.string.commonDaytimeEvening),
            subtitle = stringResource(R.string.commonDaytimeEveningRange),
            count = evening,
            color = HabitColor,
            icon = Icons.Filled.NightlightRound,
        ),
        TimeOfDayBucket(
            label = stringResource(R.string.commonDaytimeNight),
            subtitle = stringResource(R.string.commonDaytimeNightRange),
            count = night,
            color = UserColor,
            icon = Icons.Outlined.NightsStay,
        ),
    )

    Column(modifier = Modifier.fillMaxWidth()) {
        Row(verticalAlignment = Alignment.Top) {
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = stringResource(R.string.habitStatsLabelCompletionsTime),
                    fontSize = AppTheme.fsHeadingSmall,
                    fontWeight = AppTheme.fwBold,
                )
                Spacer(modifier = Modifier.height(4.dp))
                Text(
                    text = stringResource(R.string.habitStatsLabelCompletionsDistribution),
                    fontSize = AppTheme.fsBodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Spacer(modifier = Modifier.width(AppTheme.spaceLarge))
            Text(
                text = quantityString(
                    R.plurals.nHabitsCount,
                    total,
                    zeroRes = R.string.nHabitsCountZero,
                    total
                ),
                fontSize = 15.sp,
                fontWeight = AppTheme.fwExtraBold,
            )
        }

        Spacer(modifier = Modifier.height(18.dp))

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(16.dp)
                // Solid neutral track. When total is 0 the forEach below draws nothing and this
                // flat fill alone reads as the "no data yet" state - no need for a second,
                // alpha-tinted fill layered on top of an already-solid background.
                .background(
                    MaterialTheme.colorScheme.surfaceContainerHighest,
                    RoundedCornerShape(999.dp)
                ),
        ) {
            buckets.filter { it.count > 0 }.forEach { bucket ->
                Box(
                    modifier = Modifier
                        .weight(bucket.count.toFloat())
                        .fillMaxHeight()
                        .background(bucket.color),
                )
            }
        }

        Spacer(modifier = Modifier.height(18.dp))

        FlowRow(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            buckets.forEach { bucket ->
                val percent =
                    if (total == 0) 0 else ((bucket.count.toFloat() / total) * 100).roundToInt()
                // Reuses NeoCard (solid fill, accent-colored border per bucket) instead
                // of the alpha-tinted fill/border this used to have - same treatment as StatCard's
                // accent-bordered summary cards, and drops the one-off 18dp radius in favor of the
                // app-wide brNeo scale.
                NeoCard(
                    modifier = Modifier.fillMaxWidth(0.47f),
                    borderColor = bucket.color,
                    containerColor = MaterialTheme.colorScheme.surfaceVariant,
                ) {
                    Row(
                        modifier = Modifier.padding(AppTheme.spaceMedium),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        AccentBadge(color = bucket.color, size = 36.dp) {
                            Icon(
                                bucket.icon,
                                contentDescription = null,
                                tint = bucket.color,
                                modifier = Modifier.size(AppTheme.iconSizeMedium)
                            )
                        }
                        Spacer(modifier = Modifier.width(12.dp))
                        Column(modifier = Modifier.weight(1f)) {
                            Text(text = bucket.label, fontSize = 14.sp, fontWeight = AppTheme.fwBold)
                            Spacer(modifier = Modifier.height(2.dp))
                            Text(
                                text = bucket.subtitle,
                                fontSize = 11.sp,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                        Column(horizontalAlignment = Alignment.End) {
                            Text(
                                text = bucket.count.toString(),
                                fontSize = 16.sp,
                                fontWeight = AppTheme.fwExtraBold
                            )
                            Text(
                                text = "$percent%",
                                fontSize = 11.sp,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                    }
                }
            }
        }
    }
}
